### PR TITLE
Rename "auto inserting blocks" to "block hooks"

### PR DESCRIPTION
**Pull request by ockham**
## What?
Rename "auto inserting blocks" to "block hooks".

Additionally, this PR contains two minor, unrelated changes:
- [Fix a REST API schema description](https://github.com/WordPress/gutenberg/pull/54147/files#r1315078729).
- Remove a TODO note that's obsolete (as of #54029): https://github.com/WordPress/gutenberg/pull/54147/commits/8a5ed7aa0223d434242d532c3d37f4253e4ee782

## Why?
See https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1695915874:

> I've seen anecdotal feedback that `autoInsert` is not the clearest of descriptions. I'd like to propose renaming to the more familiar `hooks` terminology—and "block hooks" in more general terms—to help folks understand the mechanics and purpose more rapidly.

## How?
Replacing strings, of course. Furthermore, by hopefully picking the right replacements for different nouns and verbs previously used for auto-inserting blocks with their block hook counterpart.

For one, while [_hooks_](https://developer.wordpress.org/plugins/hooks/) is a familiar concept in WordPress, it's not really used in Core's functions names (we have `add_filter` and `add_action` but not `add_hook`); we might want to consider this for our block hook related function names.

For example, should `gutenberg_register_auto_inserted_block` become `gutenberg_add_block_hook`? That seems misleading, like it would actually add the block _hook_ -- i.e. the location where it is inserted -- rather than the block. So maybe just `gutenberg_add_block`? That will become `add_block_hook` in Core 😕 I ended up using `gutenberg_register_hooked_block` in this case.

Or: in `block.json`, should the field be called `blockHooks` or just `hooks` -- since implicitly, `block.json` is always about blocks?

## Testing Instructions
**Note that you have to re-enable the experiment, as its name has changed as well!**

You can use [v0.3.1](https://github.com/ockham/like-button/releases/download/v0.3.1/like-button.zip) of the Like Button block plugin for testing (it's been updated to use `__experimentalBlockHooks` instead of `__experimentalAutoInsert`).

Verify that ~auto-inserting blocks~ block hooks work as before.

### CODE REVIEW

**Comment by gziolo** [comment]
I agree there are some things to consider with the revised name would be whether it should be `blockHooks` or just `hooks` in `block.json`. I'm more in favor of `hooks` if is won't be too confusing. The other aspect is how we convey the same meaning with a more general name:

**Before**:

```json
"__experimentalAutoInsert": {
	"core/comment-template": "lastChild"
}
```

**After**:

```json
"hooks": {
	"core/comment-template": "lastChild"
}
```

Would it be better to go with something more detailed and should we make the hook type the key? Example:

```json
"hooks": {
	"autoInsertLastChild": "core/comment-template"
}
```

---

Other than that, I think this PR isn't complex so we can always split the decision process into two steps as soon as we agree on the UI aspect, and cover the rest in a follow-up.

**Comment by ockham** [comment]
> Would it be better to go with something more detailed and should we make the hook type the key? Example:
>
> ```json
> "hooks": {
> 	"autoInsertLastChild": "core/comment-template"
> }
> ```

I considered something like this when first working on #51449 but decided against it. The main reason was that I thought it's possible (and not even entirely implausible) for a block to be auto-inserted next to various different blocks; e.g. a Like block as a Comment Template's last block, and after Post Content:

```json
"__experimentalAutoInsert": {
	"core/comment-template": "lastChild",
	"core/post-content": "after"

}
```

Now we could of course achieve the same with the key and value flipped, if we allow arrays as values:

```json
"__experimentalAutoInsert": {
	"lastChild": "core/comment-template",
	"after": [ "core/post-content", "core/other-block" ]

}
```

... but that's more complex, so I thought why bother 😬 

---

Note that the opposite scenario -- auto-inserting a block in various different relative positions next to the _same_ block -- seems unlikely enough that I don't think we'll need to consider it:

```json
"__experimentalAutoInsert": {
	"core/post-content": [ "before", "after" ]
}
```

File: lib/experimental/block-hooks.php [hunk 1]
```diff
@@ -325,17 +330,17 @@ function gutenberg_serialize_blocks( $blocks ) {
...
-function gutenberg_register_auto_insert_rest_field() {
+function gutenberg_register_block_hooks_rest_field() {
 	register_rest_field(
 		'block-type',
-		'auto_insert',
+		'block_hooks',
 		array(
 			'schema' => array(
-				'description'       => __( 'Block types that may be automatically inserted near this block and the associated relative position where they are inserted.', 'gutenberg' ),
+				'description'       => __( 'This block is automatically inserted near any occurence of the block types used as keys of this map, into a relative position given by the corresponding value.', 'gutenberg' ),
```
Comment (ockham) [inline lib/experimental/block-hooks.php:343]: See https://github.com/WordPress/gutenberg/pull/52969#discussion_r1315074462:

> BTW while working on #54147, I discovered that my original description was inaccurate [...]: The field isn't for block types that are automatically inserted next to this block, but the other way around: we're inserting this block next to those block types 😬 
>
> I'll update, probably as part of #54147.

cc/ @dmsnell for wording 😊

File: lib/experimental/block-hooks.php [hunk 2]
```diff
@@ -1,110 +1,26 @@
 <?php
 /**
- * Auto-inserting blocks.
+ * Block hooks.
  *
  * @package gutenberg
  */
 
 /**
- * Return a function that auto-inserts a block next to a given "anchor" block.
```
Comment (ockham) [inline lib/experimental/block-hooks.php:9]: Note that the two helper methods that used to be located at the top of the file were moved further down.

**Comment by dmsnell** [comment]
I'm confused by the term "block hooks" as it has lost its specific meaning, which is inserting blocks in relation to others. I understand some relationship to WordPress hooks, but WordPress hooks are also "fire and forget" in that they are more about running code as an announcement of activity. If anything, auto-inserting blocks is closer in semantic to _filters_ since we're discussing modifying the content passed into the filter.

@ockham you had some use of "anchor" that I think could be a clue to a more descriptive name.

Using "block hooks" seems so undefined to me, and the kind of thing we could easily add later for more generic purposes than auto-inserting them. Can we find something that links back to familiar terminology without matching an overly-generic name with an over-specific behavior?

 - block anchor points
 - phantom blocks
 - related blocks
 - implied blocks
 - implicit blocks
 - block attachments (attachment points)
 - block snapping
 - auxiliary blocks

**Comment by gziolo** [comment]
> Would it be better to go with something more detailed and should we make the hook type the key? Example:
>
> ```json
> "hooks": {
> 	"autoInsertLastChild": "core/comment-template"
> }
> ```

I should clarify that better. My only concern was that `hooks` is a very general term, and by using it with a quite specific auto-inserting feature, we would limit options for future expansion. What I was wondering about could be represented also with the following examples:

```json
"insertHooks": {
	"core/comment-template": "lastChild"
}
```

or

```json
"hooks": {
	"insert": {
		"core/comment-template": "lastChild"
	}
}
```

I didn't refer to how the connection between blocks and the insertion place was shaped.

**Comment by ockham** [comment]
> I should clarify that better. My only concern was that `hooks` is a very general term, and by using it with a quite specific auto-inserting feature, we would limit options for future expansion. What I was wondering about could be represented also with the following examples:
> 
> ```json
> "insertHooks": {
> 	"core/comment-template": "lastChild"
> }
> ```
> 
> or
> 
> ```json
> "hooks": {
> 	"insert": {
> 		"core/comment-template": "lastChild"
> 	}
> }
> ```
> 
> I didn't refer to how the connection between blocks and the insertion place was shaped.

Ah, gotcha! My apologies, I was misreading you.

I see now what you mean. While I think it's a valid concern, I'm a bit worried about YAGNI: I'd rather not introduce e.g. a nested `hooks: { insert: {...} }` field when it's not clear if we're ever going to need anything else inside `hooks` (or -- maybe worse -- people might shoehorn another feature into hooks "because it's already there"). For this scenario, I'd go as far as to say if we end up needing something like that after all, it's probably fine to bump the API version.

A similar argument applies to a "flat" `insertHooks` field -- we'd kind of suggest that there will be `someOtherKindOfHooks` someday -- plus it feels a bit random if we don't call them "insert hooks" anywhere else in our docs and code 😅

File: packages/block-editor/src/hooks/block-hooks.js [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+				);
+				break;
+		}
+	};
+
+	return (
+		<InspectorControls>
+			<PanelBody
+				className="block-editor-hooks__block-hooks"
+				title={ __( 'Plugins' ) }
```
Comment (gziolo) [inline packages/block-editor/src/hooks/block-hooks.js:183]: Should it be renamed to "Block Hooks"?

**Comment by ockham** [comment]
@ndiego Can I ask you to look this over (from a documentation POV)? 🙌 (Plus native language background and all 😬 )
There's a few questions in the PR desc and in GH comments I left; would be curious to hear your thoughts on those 😊

File: lib/experiments-page.php [hunk 4]
```diff
@@ -104,14 +104,14 @@ function gutenberg_initialize_experiments_settings() {
 	);
 
 	add_settings_field(
-		'gutenberg-auto-inserting-blocks',
-		__( 'Auto-inserting blocks', 'gutenberg' ),
+		'gutenberg-block-hooks',
+		__( 'Block hooks', 'gutenberg' ),
```
Comment (gziolo) [inline lib/experiments-page.php:108]: An important note: remember to enable the experiment again before testing these changes. It took me a while to realize that.

**Comment by ockham** [comment]
@nerrad In case you'd also like to weigh in on nomenclature 😊 

For context, [this](https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1695915874) is what prompted this change:

> I've seen anecdotal feedback that `autoInsert` is not the clearest of descriptions. I'd like to propose renaming to the more familiar `hooks` terminology—and "block hooks" in more general terms—to help folks understand the mechanics and purpose more rapidly.

**Review by gziolo — APPROVED** [review]
I approve this PR, assuming "block hooks" is the final term. I like the changes applied to code occurrences of the old name in the variable or function names and to code comments that received a bit of clarification as part of the process. I can confirm that everything works as expected when you follow the testing instructions in the PR description.

**Comment by ockham** [comment]
> I'm confused by the term "block hooks" as it has lost its specific meaning, which is inserting blocks in relation to others. I understand some relationship to WordPress hooks, but WordPress hooks are also "fire and forget" in that they are more about running code as an announcement of activity. If anything, auto-inserting blocks is closer in semantic to _filters_ since we're discussing modifying the content passed into the filter.

Not totally sure I follow/agree 100% 😅  The [most canonical sources](https://developer.wordpress.org/plugins/hooks/) I've seen seem to define hooks as the union of actions (that's what you meant with "fire and forget"/announcement of activity?) and filters:

> There are two types of hooks: [Actions](https://developer.wordpress.org/plugins/hooks/actions/) and [Filters](https://developer.wordpress.org/plugins/hooks/filters/).

I agree however that auto-inserting blocks are closer in spirit to filters than actions (and thus, than their union). Still, they're dissimilar enough from filters to warrant different nomenclature; and since "hooks" is already used to denote the set of existing extension mechanisms, it might be okay to re-use them here. I guess my mental model would be that per their introduction, block hooks extend (heh) the set of extensions mechanisms:

> There are three types of hooks: [Actions](https://developer.wordpress.org/plugins/hooks/actions/), [Filters](https://developer.wordpress.org/plugins/hooks/filters/), and Block Hooks.

> @ockham you had some use of "anchor" that I think could be a clue to a more descriptive name.

Yeah, that one came rather naturally when I was looking for a way to describe the relevant function arg in some PHPDoc. I have admittedly been struggling a bit with the fact that I haven't been able to reconcile that name with "block hooks" -- I continue to call the function arguments "anchor blocks", since "block hook" doesn't quite seem to work there. The way I've been thinking about it is that if anything, a block hook is an insertion point defined by an anchor block plus a relative position; I've actually [added one sentence](https://github.com/WordPress/gutenberg/pull/54147/files#diff-0f2cd5747256918d42a50d36946e843198352dffcad62c382ab573c2a0c07c89R95) to that effect as part of this PR.

> Using "block hooks" seems so undefined to me, and the kind of thing we could easily add later for more generic purposes than auto-inserting them. Can we find something that links back to familiar terminology without matching an overly-generic name with an over-specific behavior?
> 
> * block anchor points
> * phantom blocks
> * related blocks
> * implied blocks
> * implicit blocks
> * block attachments (attachment points)
> * block snapping
> * auxiliary blocks

I like "block anchor points" (which is probably no surprise since it aligns with what I landed on earlier 😅 ), and also "implied/implicit blocks" (though I'm not quite sure they express strongly enough that these blocks are actually rendered and present in the editor 🤔 The others are a bit too vague IMO or have too much potential to be confused with other concepts. Based on what I wrote above, I'll throw "block insertion points" in the mix.

On the note of naming things overly generic when they have overly specific behavior: I get that concern (and have voiced a variant thereof in early discussions on the renaming), but I do think that this mechanism will indeed be one of the more generic ways to extend a block theme, so I'm not _that_ worried that its behavior is _that_ specific.

Anyway: The main reason for this PR was to get rid of the ["auto-inserting" moniker which doesn't seem to have resonated with people](https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1695915874) (maybe it sends the vibe that some blocks are forcefully inserted, and they can't do anything about it?), and to do so in time for GB 16.6 (to be released tomorrow) so that there can be some user testing of the feature before the WP 6.4 Feature Freeze (in three weeks from now). Since the name was suggested by Matías, I'm inclined to land the PR; I believe that the [planned Call for Testing](https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1695977712) could also serve as a testbed for the name of the feature, and since we still have the `__experimental` prefix, we'll still be able to make some tweaks before GB 16.7/WP 6.4.

I really appreciate your thoughts and suggestions ❤️ I'm mostly trying to bridge the gap between some urgency for changing the name in time for wider testing, and wider discussion. Maybe we can explicitly ask users how they like the name and provide some alternative suggestions as part of the Call for Testing? cc/ @annezazu

File: packages/block-editor/src/hooks/block-hooks.js (hunk 3 above)
Comment (ockham) [reply packages/block-editor/src/hooks/block-hooks.js:183]: I think I'd leave it as is. "Plugins" was originally suggested as the name of this panel when the feature was still called "Auto-inserting Blocks", so arguably, we have precedent that the panel name is disconnected from the feature's name. My understanding is that "Auto-inserting Blocks" or "Block Hooks" is developer-oriented terminology, which  we don't want to necessarily burden the user with; for them, it's enough to know that, "Hey, there's this plugin that would like to insert a block here."

**Comment by ockham** [comment]
I'll leave this open for a bit longer so folks can comment on it and will merge it tomorrow (before ~12:00 UTC) -- in time for GB 16.6 stable 😊 
cc/ @vcanales

**Comment by nerrad** [comment]
> @nerrad In case you'd also like to weigh in on nomenclature 😊
> 
> For context, https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1695915874 is what prompted this change:
> 
> > I've seen anecdotal feedback that autoInsert is not the clearest of descriptions. I'd like to propose renaming to the more familiar hooks terminology—and "block hooks" in more general terms—to help folks understand the mechanics and purpose more rapidly.

I dislike "block_hooks":
- It's not immediately apparent what the purpose of this setting/functionality is from the name. 
- "hooks" is a generic term that can mean any number of things, yet the API in its current implementation has a specific purpose attached to it. 
- `block.json` appears to have an architectural direction that is intended to be non-WP specific so that presumably consumers of blocks would be able to infer meaning outside of the WP context. How would this meaning be inferred from the "hook" nomenclature.

I would like to see something that is more clear - if there is reluctance to use "auto insert", then perhaps something like `inject_blocks` or to borrow from @dmsnell's brainstorming `attach_to`?

**Comment by mtias** [comment]
Thanks fro the feedback @nerrad—naming things remains one of the most challenging parts of API design :) Hooks is still a familiar term in WP parlance to signify "attach to" and it still feels adequate beyond WP. It somehow resembles the little hooks lego pieces have to combine with each other (I believe they call them _studs_). There might be a better name we can settle on during testing, but I want to avoid shipping with auto-insert for the initial round.

**Comment by ockham** [comment]
Okay, thanks all for weighing in! Going to merge this so it'll make it into GB 16.6, as originally planned 😊

**Comment by ockham** [comment]
Cherry-picked to the `release/16.6` branch in d62508d704dc1b41c9e73b69c869b75d78556dd4, and changed the milestone to Gutenberg 16.6.
cc/ @vcanales

**Comment by ZebulanStanphill** [comment]
"Block hooks" sounds awfully vague and liable to be confused with a number of existing things like the block editor PHP hooks or the React hooks used by blocks.

Additionally, it seems weird to leave "insert" entirely out of the naming, since that's all this feature is about.

Why not call it "default insertion points" or "default insertions"? That would avoid the negative connotations of "auto" ("default" implies it can be changed) while also clarifying the "when" of the block insertion.

**Pull request merged**